### PR TITLE
Support conditionally deducting from available limits

### PR DESCRIPTION
**Pull request by alisaifee**
Addresses #234

#### Only count non 200 responses towards the limit

In the example below every time the route is accessed in error a deduction will occur and when the limit is hit subsequent requests within the second will be blocked.

```python
def deduct_on_error(response):
    if response.status_code != 200:
         return True
    return False

@app.route("....")
@limiter.limit("10/second", deduct_when=deduct_on_error)
def route():
   ....

```

#### Use separate limits for errors
In this example both successful and unsuccessful requests will count towards the corresponding rate limits. If the requests all succeed there can be up to 100 requests per second, however the first unsuccessful request will result in all subsequent requests within the second being blocked.
```python

@app.route("...")
@limit.limit("1/second", deduct_when=lambda r: r.status_code != 200)
@limit.limit("100/second", deduct_when=lambda r: r.status_code == 200)
def route():
    ...
```

### CODE REVIEW

File: flask_limiter/extension.py [hunk 1]
```diff
@@ -320,7 +320,16 @@ def limiter(self):
         else:
             return self._limiter
 
+
+    def __check_conditional_deductions(self, response):
+        for lim, args in getattr(g, 'conditional_deductions', {}).items():
+            if lim.deduct_when(response):
+                if self.limiter.hit(lim.limit, *args):
+                    g.view_rate_limit = lim
```
Comment (CedricCabessa) [inline flask_limiter/extension.py:328]: `g.view_rate_limit = [lim.limit] + args`

Comment (alisaifee) [reply flask_limiter/extension.py:328]: Thanks! There's a few bugs in that one block actually 🤦

**Comment by alisaifee** [comment]
> I found one small issue :-)
> 
> It is exactly what I need 😍 can't wait for the release :-)
> 
> Do you think you can add `deduct_when` at the global `Limiter` level? So it can be applied on every route!

I thought about it and initially decided not to since it would only satisfy the first example and break down if you wanted to have two separate limits as in the second example. To maintain some consistency however, I'll add it as `default_limits_deduct_when` and additionally also add `default_limits_exempt_when` as arguments to the constructor.

The design decision of supplying default limits as constructor arguments breaks down for being able to specify different limits for different scenarios for all routes. I'll try to address this separately by allowing wrapping the entire application object with a `limit` or `shared_limit`. Similar to how it is possible to do this with `flask.Blueprint` objects ([ref](https://flask-limiter.readthedocs.io/en/stable/#rate-limiting-all-routes-in-a-flask-blueprint))

**Comment by CedricCabessa** [comment]
Just tested the new version, looks awesome :+1:

**Comment by alisaifee** [comment]
> Just tested the new version, looks awesome 👍

That's great!  Thanks for helping test & review the change. I'll release this over the weekend.

**Comment by alisaifee** [comment]
@CedricCabessa this is now released in [1.3](https://pypi.org/project/Flask-Limiter/1.3/)

**Pull request merged**